### PR TITLE
Feature/log paginator refactor

### PR DESCRIPTION
**Pull request by chennan47**
<b>Purpose</b>
change the paginators for logs from the old "more logs" to the new page style

<b>Changes</b>
the paginators for logs used to be
screen shot 2015-03-19 at 5 56 26 pm

now it has changed to actual page structure
screen shot 2015-03-19 at 5 57 19 pm

### CODE REVIEW

**Comment by chennan47** [comment]
@sloria rdy for review

File: website/static/js/paginator.js [hunk 1]
```diff
@@ -112,14 +112,14 @@ var Paginator = oop.defclass({
...
+        this.fetchResult();
     },
     previousPage: function(){
         this.currentPage(this.currentPage() - 1);
-        this.search();
+        this.fetchResult();
     },
-    search: function() {
-        throw new Error('Paginator subclass must define a "search" method.');
+    fetchResult: function() {
```
Comment (sloria) [inline website/static/js/paginator.js]: How about `fetchResults`?

File: website/views.py [hunk 2]
```diff
@@ -283,22 +284,21 @@ def dashboard(auth):
 @must_be_logged_in
 def watched_logs_get(**kwargs):
     user = kwargs['auth'].user
-    page_num = int(request.args.get('pageNum', '').strip('/') or 0)
-    page_size = 10
-    offset = page_num * page_size
-    recent_log_ids = itertools.islice(user.get_recent_log_ids(), offset, offset + page_size + 1)
+    page = int(request.args.get('page', '').strip('/') or 0)
+    size = float(request.args.get('size', 10))
```
Comment (sloria) [inline website/views.py]: Shouldn't `size` be an integer?

File: website/views.py [hunk 3]
```diff
@@ -283,22 +284,21 @@ def dashboard(auth):
 @must_be_logged_in
 def watched_logs_get(**kwargs):
     user = kwargs['auth'].user
-    page_num = int(request.args.get('pageNum', '').strip('/') or 0)
-    page_size = 10
-    offset = page_num * page_size
-    recent_log_ids = itertools.islice(user.get_recent_log_ids(), offset, offset + page_size + 1)
+    page = int(request.args.get('page', '').strip('/') or 0)
```
Comment (sloria) [inline website/views.py]: What if the client passes in invalid data? Invalid inputs should be handled by returning the proper error response to the user.

File: website/static/css/site.css [hunk 4]
```diff
@@ -1946,6 +1946,10 @@ span#profileFullname{
     padding-bottom: 20px;
 }
 
+/*Dashboard */
+.col-sm-5 .tab-content .addon-widget-body{
```
Comment (sloria) [inline website/static/css/site.css]: This is a brittle selector; what happens if we change, e.g. from `.col-sm-5` to `.col-sm-4`?

**Comment by sloria** [comment]
1st pass finished. :point_up:

File: website/static/js/paginator.js (hunk 1 above)
Comment (chennan47) [inline website/static/js/paginator.js]: solved in  5c4e8b8

File: website/views.py (hunk 2 above)
Comment (chennan47) [inline website/views.py]: solved in 724e95f

File: website/static/css/site.css (hunk 4 above)
Comment (chennan47) [inline website/static/css/site.css]: solved in 2cd8abf

**Comment by chennan47** [comment]
@sloria rdy for the 2nd pass

File: website/views.py [hunk 5]
```diff
@@ -283,22 +284,32 @@ def dashboard(auth):
...
     user = kwargs['auth'].user
-    page_num = int(request.args.get('pageNum', '').strip('/') or 0)
-    page_size = 10
-    offset = page_num * page_size
-    recent_log_ids = itertools.islice(user.get_recent_log_ids(), offset, offset + page_size + 1)
+    try:
+        page = int(request.args.get('page', 0))
+    except ValueError:
+        raise HTTPError(http.BAD_REQUEST, data=dict(
+            message_long='Invalid value for "page": {}'.format(request.args['page'])
```
Comment (sloria) [inline website/views.py]: To be safe, better not to return the user's input (which may be unsafe) in the response.

File: website/views.py [hunk 6]
```diff
@@ -283,22 +284,32 @@ def dashboard(auth):
...
+        ))
+    try:
+        size = int(request.args.get('size', 10))
+    except ValueError:
+        raise HTTPError(http.BAD_REQUEST, data=dict(
+            message_long='Invalid value for "size": {}'.format(request.args['size'])
```
Comment (sloria) [inline website/views.py]: See comment above.

File: website/views.py [hunk 7]
```diff
@@ -283,22 +284,32 @@ def dashboard(auth):
 @must_be_logged_in
 def watched_logs_get(**kwargs):
     user = kwargs['auth'].user
-    page_num = int(request.args.get('pageNum', '').strip('/') or 0)
-    page_size = 10
-    offset = page_num * page_size
-    recent_log_ids = itertools.islice(user.get_recent_log_ids(), offset, offset + page_size + 1)
+    try:
+        page = int(request.args.get('page', 0))
+    except ValueError:
```
Comment (sloria) [inline website/views.py]: A `TypeError` might also be raised here, if `None` is the user input.

File: tests/test_views.py [hunk 8]
```diff
@@ -517,10 +541,12 @@ def test_get_logs_with_count_param(self):
                 )
             )
         self.project.save()
-        url = '/api/v1/project/{0}/log/'.format(self.project._primary_key)
+        url = self.project.api_url_for('get_logs')
         res = self.app.get(url, {'count': 3}, auth=self.auth)
         assert_equal(len(res.json['logs']), 3)
-        assert_true(res.json['has_more_logs'])
+        assert_equal(res.json['total'], 5 + 2)
```
Comment (sloria) [inline tests/test_views.py]: Why `+ 2`? Can you add a code comment?

File: tests/test_views.py [hunk 9]
```diff
@@ -1949,11 +1998,54 @@ def test_get_more_watched_logs(self):
...
         assert_equal(len(res.json['logs']), 3)
+        assert_equal(res.json['total'], 12 + 1)
+        assert_equal(res.json['page'], page)
+        assert_equal(res.json['pages'], 2)
         assert_equal(res.json['logs'][0]['action'], 'file_added')
 
+    def test_get_more_watched_logs_invalid_page(self):
+        project = ProjectFactory()
+        # Add some logs
+        for _ in range(12):
```
Comment (sloria) [inline tests/test_views.py]: Is this setup even necessary if the request is invalid anyway?

File: website/views.py (hunk 7 above)
Comment (sloria) [inline website/views.py]: Disregard. I don't think there's a ways for the user to pass `None` in the querystring.

File: website/views.py (hunk 5 above)
Comment (chennan47) [inline website/views.py]: solved in  996f4f3

File: website/views.py (hunk 6 above)
Comment (chennan47) [inline website/views.py]: solved in  996f4f3

File: tests/test_views.py (hunk 9 above)
Comment (chennan47) [inline tests/test_views.py]: solved in  93bfd41

File: tests/test_views.py (hunk 8 above)
Comment (chennan47) [inline tests/test_views.py]: solved in 4c7c2cd

**Comment by chennan47** [comment]
@sloria rdy for another pass.

File: website/views.py [hunk 10]
```diff
@@ -283,22 +284,32 @@ def dashboard(auth):
...
+    try:
+        size = int(request.args.get('size', 10))
+    except ValueError:
+        raise HTTPError(http.BAD_REQUEST, data=dict(
+            message_long='Invalid value for "size".'
+        ))
+
+    start = page * size
+    total = sum(1 for x in user.get_recent_log_ids())
+    recent_log_ids = itertools.islice(user.get_recent_log_ids(), start, start + size)
```
Comment (sloria) [inline website/views.py]: This is duplicated with the contributor pagination logic? Can we DRY this up somehow?

Comment (chennan47) [inline website/views.py]: solved in  fda1825

**Comment by chennan47** [comment]
@sloria rdy for another pass

File: website/static/js/logFeed.js [hunk 11]
```diff
@@ -75,47 +77,51 @@ var Log = function(params) {
...
-var LogsViewModel = function(logs, hasMoreLogs, url) {
-    var self = this;
-    self.enableMoreLogs = ko.observable(hasMoreLogs);
-    self.logs = ko.observableArray(logs);
-    var pageNum=  0;
-    self.url = url;
-
+var LogsViewModel = oop.extend(Paginator, {
+    constructor: function(logs, url) {
+        this.super.constructor();
```
Comment (sloria) [inline website/static/js/logFeed.js]: This needs to be

```js
this.super.constructor.call(this);
```

to ensure that the `this` object is correct.

File: website/static/js/logFeed.js [hunk 12]
```diff
@@ -75,47 +77,51 @@ var Log = function(params) {
...
+        var self = this;
+        self.logs = ko.observableArray(logs);
+        self.url = url;
+
+        self.tzname = ko.computed(function() {
```
Comment (sloria) [inline website/static/js/logFeed.js]: This can be a `pureComputed`.

**Comment by sloria** [comment]
Pass finished. Just minor changes. :ocean:

File: website/static/js/logFeed.js (hunk 11 above)
Comment (chennan47) [inline website/static/js/logFeed.js]: solved in 70a2285

File: website/static/js/logFeed.js (hunk 12 above)
Comment (chennan47) [inline website/static/js/logFeed.js]: solved in 63c4695

**Comment by chennan47** [comment]
@sloria minor changes have been made. rdy for review and merge.

**Comment by sloria** [comment]
@chennan47 Thanks. How difficult would it be to make the paginator stationary within the panel container. One annoyance is that the pagination controls move around while you're paging (depending on line breaks and number of logs).

**Comment by lbanner** [comment]
Yeah -- I see that too -- Anchoring it is a good idea ... +1

**Comment by GaryKriebel** [comment]
I showed @chennan47 how the controls move in the logs.  He was going to discuss with @sloria issues with implementation.

**Comment by chennan47** [comment]
@sloria @lbanner @GaryKriebel 
after discussion, the current solution is setting the paginator bar at the bottom of the page.
screen shot 2015-03-26 at 4 43 45 pm
screen shot 2015-03-26 at 4 44 07 pm

it still has some side effect that if the first pages' log is super long that exceed the min-height set for the  panel, but on the second page there is no long logs or less than 10 logs that the paginator bar will still move slightly. But this will be quite the edge case and we don't have a better solution for now.

**Comment by chennan47** [comment]
@sloria the css change has already been made. rdy for review.

**Comment by lbanner** [comment]
I see what you mean ... there is a little bit of movement based on # items in log for whole frame it seems ... but the anchor to the bottom is much better.

**Comment by sloria** [comment]
@chennan47 Thanks. much better. :octopus:

**Pull request merged**